cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,357 @@
+      Set<EndPointField> endPointFields = GSON.fromJson(value, SET_ENDPOINT_FIELD_TYPE);
+      if (endPointFields != null) {
+        // TODO: can it be null ever??
+        result.addAll(endPointFields);
+      }
+    }
+
+    return result;
+  }
+
+  public Set<ProgramRunOperations> getIncomingOperations(EndPoint endPoint, long start, long end) {
+    return getOperations(INCOMING_DIRECTION_MARKER, endPoint, start, end);
+  }
+
+  public Set<ProgramRunOperations> getOutgoingOperations(EndPoint endPoint, long start, long end) {
+    return getOperations(OUTGOING_DIRECTION_MARKER, endPoint, start, end);
+  }
+
+  private Set<ProgramRunOperations> getOperations(byte[] direction, EndPoint endPoint, long start, long end) {
+    Map<Long, Set<ProgramRunId>> checksumsWithProgramRunsInRange
+            = getChecksumsWithProgramRunsInRange(direction, endPoint, start, end);
+
+    Set<Long> checksums = checksumsWithProgramRunsInRange.keySet();
+    Set<ProgramRunOperations> result = new HashSet<>();
+
+    for (long checksum : checksums) {
+      byte[] rowKey = getChecksumRowKey(checksum);
+      String value = Bytes.toString(table.get(rowKey, RAW_OPERATION_MARKER));
+      Set<Operation> operations = GSON.fromJson(value, SET_OPERATION_TYPE);
+      if (operations != null) {
+        result.add(new ProgramRunOperations(checksum, checksumsWithProgramRunsInRange.get(checksum), operations));
+      }
+    }
+
+    return result;
+  }
+
+  private Map<Long, Set<ProgramRunId>> getChecksumsWithProgramRunsInRange(byte[] direction, EndPoint endPoint,
+                                                                          long start, long end) {
+    byte[] scanStartKey = getScanStartKey(direction, endPoint, end);
+    byte[] scanEndKey = getScanEndKey(direction, endPoint, start);
+    Map<Long, Set<ProgramRunId>> result = new HashMap<>();
+    try (Scanner scanner = table.scan(scanStartKey, scanEndKey)) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("Got row key = {}", Bytes.toString(row.getRow()));
+        }
+
+        long checksum = Bytes.toLong(row.get(CHECKSUM_MARKER));
+        ProgramRunId programRunId = GSON.fromJson(Bytes.toString(row.get(PROGRAM_MARKER)), ProgramRunId.class);
+        Set<ProgramRunId> programRuns = result.computeIfAbsent(checksum, k -> new HashSet<>());
+        programRuns.add(programRunId);
+      }
+    }
+    return result;
+  }
+
+  private byte[] getScanStartKey(byte[] direction, EndPoint endPoint, long end) {
+    // time is inverted, hence we need to have end time in start key.
+    // Since end time is exclusive, add 1 to make it inclusive.
+    return getScanKey(direction, endPoint, end + 1);
+  }
+
+  private byte[] getScanEndKey(byte[] direction, EndPoint endPoint, long start) {
+    // time is inverted, hence we need to have start time in end key.
+    // Since start time is inclusive, subtract 1 to make it exclusive.
```

Comment:
removed these methods.. we only need to add 1 always to the inverted time. the reason is
once the time is inverted, original end-time becomes the start time for the hbase scan, however we should not return the record corresponding to this start time (inverted end-time) since end time is exclusive, so we add 1. Similarly once the time is inverted, original start-time becomes the end time of the scan. Since we want to return the records corresponding to this end time (inverted start-time), we add 1.